internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,190 @@
+	}
+
+	batch.Add(fmt.Errorf("%w %v", errNotFound, missingMetrics(metricsUnderTest)))
+
+	return batch
+}
+
+func missingMetrics(metrics map[string]struct{}) []string {
+	missing := []string{}
+	if len(metrics) > 0 {
+		for m := range metrics {
+			missing = append(missing, m)
+		}
+	}
+	return missing
+}
+
+// buildMetricNames creates a set of all the expected metrics names for the given prefix
+// that should be registered with Prometheus as defined in the baseplate spec.
+// The prefix will be either thrift, http, or grpc.
+func buildMetricNames(prefix string) map[string]struct{} {
+	suffixes := []string{"latency_seconds", "requests_total", "active_requests"}
+	names := map[string]struct{}{}
+	for _, suffix := range suffixes {
+		names[prometheus.BuildFQName(prefix, server, suffix)] = struct{}{}
+		names[prometheus.BuildFQName(prefix, client, suffix)] = struct{}{}
+	}
+	return names
+}
+
+// validateName checks the following:
+// 1) the metric name has at least 3 parts separated by "_".
+// The parts are <namespace>_<client/server>_<name>_<suffix>, where suffix is optional.
+// Ref: https://prometheus.io/docs/practices/naming
+// 2) the metric name has the correct prefix.
+// 3) the metric contains either "client" or "server" as the second part.
+func validateName(name, prefix string) errorsbp.Batch {
+	var batch errorsbp.Batch
+
+	const (
+		separator = "_"
+		partCount = 3
+	)
+	parts := strings.SplitN(name, separator, partCount)
+	if len(parts) < partCount {
+		batch.Add(fmt.Errorf("%w metric %q does not have minimum number (%d) of parts", errLength, name, partCount))
+		return batch
+	}
+
+	if got, want := parts[0], prefix; got != want {
+		batch.Add(fmt.Errorf("%w metric %q does not have the expected prefix %q", errPrefix, name, prefix))
+	}
+
+	if parts[1] != client && parts[1] != server {
+		batch.Add(fmt.Errorf("%w metric %q", errClientServer, name))
+	}
+	return batch
+}
+
+func validateLabels(name, prefix string, gotLabels map[string]struct{}) errorsbp.Batch {
+	var batch errorsbp.Batch
+
+	wantLabels := buildLables(name, prefix)
+	if diff := cmp.Diff(gotLabels, wantLabels); diff != "" {
+		batch.Add(fmt.Errorf("%w: (-got +want)\n%s", errDiffLabels, diff))
+	}
+
+	return batch
+}
+
+// buildLables returns a set of expected labels for the metric name provided.
+// prefix is either thrift, http, or grpc.
+// latency_seconds metrics expect the following labels:
+//   - "<prefix>_service"
+//   - "<prefix>_method"
+//   - "<prefix>_success"
+// request_total metrics expect the following labels:
+//   - "<prefix>_service"
+//   - "<prefix>_method"
+//   - "<prefix>_success"
+//   - "<prefix>_exception_type"
+//   - "<prefix>_baseplate_status"
+//   - "<prefix>_baseplate_status_code"
+// active_requests metrics expect the following labels:
+//   - "<prefix>_service"
+//   - "<prefix>_method"
+func buildLables(name, prefix string) map[string]struct{} {
+	labelSuffixes := []string{"service", "method"}
+	successLabelSuffix := "success"
```

Comment:
🔕 factoring this out to avoid the same literal twice makes it a bit jarring on line 167 to switch from a var to literals; I'd just use literals the two places this appears